lldb/tools/lldb-server/lldb-platform.cpp

```diff
@@ -393,125 +362,222 @@ int main_platform(int argc, char *argv[]) {
-  // The highest this should get reasonably is a function of the number
-  // of target CPUs. For now, let's just use 100.
-  const int backlog = 100;
+  if (gdbserver_port != 0 &&
+      (gdbserver_port < LOW_PORT || gdbserver_port > HIGH_PORT)) {
+    WithColor::error() << llvm::formatv("Port number {0} is not in the "
+                                        "valid user port range of {1} - {2}\n",
+                                        gdbserver_port, LOW_PORT, HIGH_PORT);
+    return 1;
+  }
 
-  std::unique_ptr<Acceptor> acceptor_up(Acceptor::Create(
-      listen_host_port, children_inherit_listen_socket, error));
-  if (error.Fail()) {
-    fprintf(stderr, "failed to create acceptor: %s", error.AsCString());
-    exit(socket_error);
+  std::string hostname;
+  uint16_t platform_port = 0;
+
+  // Try to match socket name as URL - e.g., tcp://localhost:5555
+  if (std::optional<URI> uri = URI::Parse(listen_host_port)) {
+    if (!Socket::FindProtocolByScheme(uri->scheme.str().c_str(), protocol)) {
+      fprintf(stderr, "Unknown protocol scheme \"%s\".",
+              uri->scheme.str().c_str());
+      return socket_error;
+    }
+    if (protocol == Socket::ProtocolTcp) {
+      hostname = uri->hostname;
+      if (uri->port) {
+        platform_port = *(uri->port);
+      }
+    } else
+      hostname = listen_host_port.substr(uri->scheme.size() + strlen("://"));
+  } else {
+    // Try to match socket name as $host:port - e.g., localhost:5555
+    llvm::Expected<Socket::HostAndPort> host_port =
+        Socket::DecodeHostAndPort(listen_host_port);
+    if (!llvm::errorToBool(host_port.takeError())) {
+      protocol = Socket::ProtocolTcp;
+      hostname = host_port->hostname;
+      platform_port = host_port->port;
+    } else
+      hostname = listen_host_port;
   }
 
-  error = acceptor_up->Listen(backlog);
+  if (protocol == Socket::ProtocolTcp) {
+    if (platform_port != 0 && platform_port == gdbserver_port) {
+      fprintf(stderr, "The same platform and gdb ports %u.", platform_port);
+      return socket_error;
+    }
+  } else {
+    if (gdbserver_port) {
+      fprintf(stderr,
+              "--gdbserver-port %u is redundant for non-tcp protocol %s.",
+              gdbserver_port, Socket::FindSchemeByProtocol(protocol));
+      return socket_error;
+    }
+    if (g_server) {
+      fprintf(stderr,
+              "Ambiguous parameters --server --listen %s.\n"
+              "The protocol must be tcp for the server mode.",
+              listen_host_port.c_str());
+      return socket_error;
+    }
+  }
+
+  std::unique_ptr<Socket> sock_platform =
+      Socket::Create(protocol, /*child_processes_inherit=*/false, error);
+  if (error.Fail()) {
+    printf("Failed to create platform socket: %s\n", error.AsCString());
+    return socket_error;
+  }
+  error = sock_platform->Listen(
+      protocol == Socket::ProtocolTcp
+          ? llvm::formatv("{0}:{1}", hostname, platform_port).str()
+          : hostname,
```

Comment:
Is `hostname` the unix socket path? That's definitely surprising. Can you instead define a single (string) variable called address (or whatever is the name of the argument to `Socket::Listen`, and assign it the appropriate value in the if block on line 415?

Reply:
Done